rclpy/src/rclpy/context.cpp

```diff
@@ -45,4 +47,76 @@ context_get_domain_id(py::capsule pycontext)
 
   return domain_id;
 }
+
+void
+_rclpy_context_handle_destructor(void * p)
+{
+  auto context = static_cast<rcl_context_t *>(p);
+  if (!context) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "_rclpy_context_handle_destructor failed to get pointer");
+    return;
+  }
+  if (NULL != context->impl) {
+    rcl_ret_t ret;
+    if (rcl_context_is_valid(context)) {
+      // shutdown first, if still valid
+      ret = rcl_shutdown(context);
+      if (RCL_RET_OK != ret) {
+        fprintf(
+          stderr,
+          "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+          "failed to shutdown rcl_context_t (%d) during PyCapsule destructor: %s\n",
+          ret,
+          rcl_get_error_string().str);
+        rcl_reset_error();
+      }
+    }
+    ret = rcl_context_fini(context);
+    if (RCL_RET_OK != ret) {
+      fprintf(
+        stderr,
+        "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+        "failed to fini rcl_context_t (%d) during PyCapsule destructor: %s\n",
+        ret,
+        rcl_get_error_string().str);
+      rcl_reset_error();
+    }
+  }
+  PyMem_FREE(context);
+}
+
+py::capsule
+create_context()
+{
+  auto context = static_cast<rcl_context_t *>(PyMem_Malloc(sizeof(rcl_context_t)));
+  if (!context) {
+    throw std::bad_alloc();
+  }
+  *context = rcl_get_zero_initialized_context();
+  PyObject * capsule = rclpy_create_handle_capsule(
```

Comment:
Totally out of scope for this PR, but would it make sense to make this function throw the exception once we're all moved over to Pybind11?

Reply:
As in, would it make sense to replace `rclpy_create_handle_capsule()` with something that throws an exception? If I understand the question correctly, once everything's converted to Pybind11 I think it would make sense to switch over to C++ classes since classes will replace the capsules altogether.